api/plgd/plgd_time.c

```diff
@@ -0,0 +1,562 @@
+/****************************************************************************
+ *
+ * Copyright 2023 Daniel Adam, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"),
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+#include "util/oc_features.h"
+
+#ifdef OC_HAS_FEATURE_PLGD_TIME
+
+#include "plgd_time_internal.h"
+#include "api/oc_core_res_internal.h"
+#include "api/oc_rep_internal.h"
+#include "api/oc_resource_internal.h"
+#include "api/oc_ri_internal.h"
+#include "api/oc_server_api_internal.h"
+#include "api/oc_storage_internal.h"
+#include "oc_api.h"
+#include "oc_core_res.h"
+#include "oc_ri.h"
+#include "port/oc_clock.h"
+#include "port/oc_log.h"
+#include "util/oc_compiler.h"
+#include "util/oc_macros.h"
+
+#ifdef OC_SECURITY
+#include "security/oc_pstat.h"
+#include "security/oc_security_internal.h"
+#endif /* OC_SECURITY */
+
+#include <inttypes.h>
+#include <math.h>
+#include <string.h>
+#include <time.h>
+
+static plgd_time_t g_oc_plgd_time = { 0 };
+
+plgd_time_t *
+plgd_time_get(void)
+{
+  return &g_oc_plgd_time;
+}
+
+static void
+dev_set_system_time(void)
+{
+  oc_clock_time_t pt = plgd_time();
+  int ret =
+    g_oc_plgd_time.set_system_time(pt, g_oc_plgd_time.set_system_time_data);
+  if (ret != 0) {
+    OC_ERR("failed to set system time: error(%d)", ret);
+    return;
+  }
+  OC_DBG("system time set to: %ld", (long)pt);
+}
+
+void
+plgd_time_set(oc_clock_time_t last_synced_time, oc_clock_time_t update_time,
+              bool dump, bool notify)
+{
+  g_oc_plgd_time.store.last_synced_time = last_synced_time;
+  g_oc_plgd_time.update_time = update_time;
+
```

Comment:
g_oc_plgd_time.status = PLGD_TIME_STATUS_IN_SYNC;
otherwise status will report when IN_SYNC_FROM_STORAGE when someone set the value and he doesn't use API for change status.

Reply:
Updated